src/input.rs

```diff
@@ -0,0 +1,497 @@
+
+    /// This method should be called in application.rs after render, so States from events will be properly updated.
+    pub fn handle_update(&mut self) {
+        self.last_mouse_position = self.mouse_position;
+        self.mouse_scroll_delta = 0.0;
+
+        self.states.retain(|_btn, state| match state {
+            State::Pressed => {
+                *state = State::Hold;
+                true
+            }
+            State::Released => false,
+            _ => true,
+        });
+    }
+
+    /// This method should be called in application.rs in event_loop, so systems will get actual Input data.
+    pub fn handle_event(&mut self, event: &winit::event::Event<'_, ()>) {
+        match event {
+            Event::WindowEvent {
+                event: WindowEvent::KeyboardInput { input, .. },
+                ..
+            } => self.handle_keyboard_event(input),
+            Event::WindowEvent {
+                event: WindowEvent::MouseInput { state, button, .. },
+                ..
+            } => self.handle_mouse_click_event(*state, *button),
+            Event::WindowEvent {
+                event: WindowEvent::CursorMoved { position, .. },
+                ..
+            } => self.handle_cursor_moved_event(position),
+            Event::WindowEvent {
+                event: WindowEvent::MouseWheel { delta, .. },
+                ..
+            } => self.handle_mouse_wheel_event(&delta),
+            Event::WindowEvent {
+                event: WindowEvent::Resized(size),
+                ..
+            } => self.window_size = Vector2::new(size.width as f32, size.height as f32),
+            _ => {}
+        }
+    }
+
+    /// Handle cursor moved event from winit.
+    pub fn handle_cursor_moved_event(&mut self, position: &winit::dpi::PhysicalPosition<f64>) {
+        self.mouse_position = Vector2 {
+            x: position.x as f32,
+            y: position.y as f32,
+        };
+    }
+
+    /// Handle keyboard event from winit.
+    pub fn handle_keyboard_event(&mut self, input: &KeyboardInput) {
+        if input.virtual_keycode.is_some() {
+            let key_code: u32 = input.virtual_keycode.unwrap() as u32;
+            match KeyCode::try_from(key_code) {
+                Ok(val) => {self.handle_button_state(Button::Key(val), input.state)},
+                Err(_) => {},
+            }
+        }
+    }
+
+    fn handle_button_state(&mut self, btn: Button, state: ElementState) {
+        match state {
+            ElementState::Pressed => {
+                if !self.states.contains_key(&btn) || State::Released == self.states[&btn] {
+                    self.states.insert(btn, State::Pressed);
+                }
+            }
+            ElementState::Released => {
+                self.states.insert(btn, State::Released);
+            }
+        }
+    }
+
+    /// Handle mouse click events from winit.
+    pub fn handle_mouse_click_event(&mut self, state: ElementState, button: winit::event::MouseButton) {
```

Comment:
rest of `handle_` functions should not be `pub`, as soon as they are used only inside of the module